proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,317 @@
+    otherwise throw `InvalidAccountOwner`
+    - the program account is at least as long enough for the header,
+    otherwise throw `AccountDataTooSmall`
+    - the authority stored in the program account is the one provided,
+    otherwise throw `IncorrectAuthority`
+    - the status stored in the program account is not finalized,
+    otherwise throw `Immutable`
+  - Copy the address of the next version into the next version field stored in
+  the previous versions program account
+  - Change the status stored in the program account to finalized
+
+## Impact
+
+This proposal:
+
+- covers all the use cases loader-v3 had but in a cleaner way and comes with
+a specification.
+- allows finalized programs to mark which other program supersedes them which
+can then be offered as an option in forntends. This provides a more secure
+alternative to redeployment / upgrading of programs at the same address.
+- makes deployment slightly cheaper for dapp developers as they would no longer
+have to burn funds for the rent exception of the proxy account.
+- provides an alternative redeployment path which does not require a big
+deposit of funds for rent exception during the upload.
+- enables dapp developers to withdrawl the surplus of funds required for rent
+exception when shortening the length of program accounts or closing them.
+- shortens the workflow of temporarily closing a program to a single
+instruction, instead of having to build and redeploy an empty program.
+- properly alignes the executable file relative to the beginning of the
+account. In loader-v3 it is misaligned.
+- once all loader-v3 programs are migrated:
+  - allows transaction account loading to be simplifed, because every program
+  would load exactly one account, no need to load the proxy account to get to
+  the actual program data (which is not listed in the transaction accounts).
+  - allows the removal of the write lock demotion exception if loader-v3 is
+  present in a transaction.
+  - corrects the miscounting of the proxy account size towards the total
+  transaction account loading limit.
+
+## Security Considerations
+
+None.
+
+## Backwards Compatibility
+
+This proposal does not break any existing programs. However, dapp developers
+might want to profit from the new program mangement instructions without
+influencing their users work flows. To do so they would need a way to turn the
+program accounts of loader-v3 to program accounts of loader-v4, changing the
+account owner but keeping the program address. A potential issue is that the
+programdata header of loader-v3 is only 45 bytes long while loader-v4 takes 48
+bytes. An automatic mechanism in the program runtime (triggered by feature
+activation) could then perform the following steps per program:
+
+- loader-v3 clears the program proxy account (setting its size to zero)
+- loader-v3 transfers all funds from the programdata to the proxy account
+- loader-v3 gifts the program proxy account to loader-v4
```

Comment:
How would the gifting work? Do we need to add a new instruction to the v3 loader? Will this upgrade mechanism be specified in a future SIMD?

Reply:
I see three options here:

1. A migrate instruction for loader-v3 which the dApp devs trigger themselves
2. A migrate instruction for loader-v3 which we the protocol engineers trigger
3. A feature gate which scans the accounts DB and migrates all loader-v3 programs

I would prefer the second option as that allows us to keep the gathering of the loader-v3 program list and then slowly working through it off-chain and still makes sure that all loader-v3 programs get migrated.